src/main/java/com/nutomic/syncthingandroid/syncthing/EventBasedModel.java

```diff
@@ -0,0 +1,272 @@
+            mChangeEventPending |= mConnections.remove(deviceId) != null;
+        } catch (JSONException e) {
+            Log.w(TAG, MessageFormat.format("Could not extract required information from data: {0}: data={1}", e.getMessage(), data), e);
+        }
+    }
+
+    private void handleDeviceConnected(JSONObject data) {
+        try {
+            String deviceId = data.getString("id");
+            RestApi.Connection c = new RestApi.Connection();
+            c.address = data.optString("addr");
+            c.clientVersion = data.optString("clientVersion");
+            c.deviceName = data.optString("deviceName");
+            c.connected = true;
+            mConnections.put(deviceId, c);
+            mChangeEventPending |= true;
+        } catch (JSONException e) {
+            Log.w(TAG, MessageFormat.format("Could not extract required information from data: {0}: data={1}", e.getMessage(), data), e);
+        } // try/catch
+    }
+
+    private void handleFolderCompletion(JSONObject data) {
+        try {
+            String deviceId = data.getString("device");
+            double completion = data.getDouble("completion");
+            String folderName = data.getString("folder");
+
+            Map<String, Double> folderCompletions = mDeviceFolderCompletion.get(deviceId);
+            if (folderCompletions == null) {
+                folderCompletions = new HashMap<>();
+                mDeviceFolderCompletion.put(deviceId, folderCompletions);
+            }
+            folderCompletions.put(folderName, completion);
+            mChangeEventPending = true;
+        } catch (JSONException e) {
+            Log.w(TAG, MessageFormat.format("Could not extract required information from data: {0}: data={1}", e.getMessage(), data), e);
+        }
+    }
+
+    private void handleStateChanged(JSONObject data) {
+        try {
+            String folder = data.getString("folder");
+            String newState = data.getString("to");
+
+            mFolderState.put(folder, newState);
+            mChangeEventPending = true;
+        } catch (JSONException e) {
+            Log.w(TAG, MessageFormat.format("Could not extract required information from data: {0}: data={1}", e.getMessage(), data), e);
+        }
+    }
+
+    public EventBasedModel.Snapshot createSnapshot() {
+        // deep-copy via serialization + deserializationt
+        try {
+            ByteArrayOutputStream bos = new ByteArrayOutputStream();
+            ObjectOutputStream oos = new ObjectOutputStream(bos);
+            oos.writeObject(this);
+            oos.close();
+
+            ByteArrayInputStream bin = new ByteArrayInputStream(bos.toByteArray());
+            ObjectInputStream oin = new ObjectInputStream(bin);
+            return new Snapshot((EventBasedModel) oin.readObject());
+        } catch (Exception e) {
+            throw new IllegalStateException("Failed to deep-copy EventBasedModel: " + e.getMessage(), e);
```

Comment:
Habit from work, where we tend to have a shorter log for tailing on a console (`info.log`) with just having the `Log Message: TheExceptionMessage` and an additional `verbose.log` including the full stack-trace. Gonna change :)